File: python/720.词典中最长的单词.py
#
# @lc app=leetcode.cn id=720 lang=python
#
# [720] 词典中最长的单词
#
# 解法1(T92% S52%)：先按照字符串长度排序，利用一个set()，一次遍历：如果单词长度为1 or 当前字符串去掉最后一位在set中，则代表该字符串的前序一定在词典中，则把当前单词也加入set并更新是否为最长的。为什么要排序呢，因为要满足一次只多一个字符，如果是只有a abc是不满足的

# @lc code=start
class Solution(object):
    def longestWord(self, words):
        """
        :type words: List[str]
        :rtype: str
        """
        words_sorted = sorted(words)
        maxStr = ""
        myset = set()
        for word in words_sorted:
            if len(word) == 1 or word[:-1] in myset:
                myset.add(word)
                if len(word) > len(maxStr): maxStr = word
        return maxStr
        
# @lc code=end

File: javascript/720.词典中最长的单词.js
/*
 * @lc app=leetcode.cn id=720 lang=javascript
 *
 * [720] 词典中最长的单词
 * 
 * 注意首先要排序，set中的key要一点一点加，不能直接就把原数组加进去，因为可能出现虽然少一位在，但少两位的可能不在，但这并不满足题目要求
 */

// @lc code=start
/**
 * @param {string[]} words
 * @return {string}
 */
var longestWord = function(words) {
    words.sort((a, b) => (a.length-b.length));
    
    var set = new Set([""]);
    var res = "";
    for (var word of words) {
        if (word.length >= res.length && set.has(word.slice(0, word.length-1))) {
            set.add(word);
            if (word.length > res.length || (word.length==res.length && word<res)) {
                res = word;
            }
        }
    }
    return res;
};
// @lc code=end

